docs/source/learn/trace-vs-execution-time.md

```diff
@@ -0,0 +1,175 @@
+
+**If you measure the duration of code that only performs XLA operations without an
+explicit instruction to wait for the device, you are primarily measuring this
+tracing time plus Python overhead. This is NOT an accurate measure of your
+model's performance on the XLA device.**
+
+## Tracing: Defining the Computation (Host-Side Activity)
+
+When you write PyTorch code using XLA tensors (e.g., tensors on a TPU),
+PyTorch/XLA doesn't execute each operation on the device right away. It traces
+these operations, adding them to an internal computation graph. If you measure
+the duration of code that only performs XLA operations without an explicit
+instruction to wait for the device, you are primarily measuring this tracing
+time plus Python overhead.
+
+### Illustrating a Common Pitfall: Measuring Only Tracing Time
+
+Consider the following conceptual code:
+
+```Python
+# Assume 'a' and 'b' are XLA tensors
+# start_time = time.perf_counter()
+
+# This operation is recorded in PyTorch/XLA's graph
+result = torch.matmul(a, b)
+
+#!!! INCORRECT PROFILING: compilation and execution are deferred !!!
+# end_time = time.perf_counter()
+# elapsed_time = end_time - start_time
+```
+
+The `elapsed_time` here would predominantly reflect how long it took PyTorch/XLA
+to trace the matmul operation. The actual matrix multiplication on the XLA
+device is not started.
+
+## Device Time: Compilation and Execution on the Accelerator
+
+To correctly profile the performance of your code on the XLA device, you must
+ensure that your timing captures the phase where the device actually does the
+work. This involves:
+
+1. Ensure the traced computational graph is sent to the device for execution.
+
+1. Make sure the Python script wait until the XLA device has completed all its
+assigned computations before taking the final timestamp.
+
+### Triggering Execution and Ensuring Completion
+
+Several mechanisms trigger graph execution and/or ensure completion:
+
+1. `torch_xla.sync(wait=True)`: This is the most direct method for benchmarking.
+It ensures all pending XLA operations are launched and, crucially, blocks the
+Python script until the device finishes.
+
+1. `Data Access/Transfer`: Operations like `tensor.cpu()`, `tensor.item()`, or
+printing an XLA tensor require the actual data. To provide it, PyTorch/XLA must
+execute the graph that produces the tensor and wait for its completion.
+
+1. `torch_xla.core.xla_model.mark_step()`: Commonly used in loops, this signals
+to XLA that a logical unit of work (e.g., a training step) is defined and can be
+executed. It triggers execution but doesn't inherently block Python until device
+completion.
+
+1. `torch_xla.core.xla_model.optimizer_step(optimizer)`: In training, this
+usually implies `mark_step` and ensures relevant computations are done.
+
+1. `torch_xla.core.xla_model.unlazy(tensors)`: Blocks until specified tensors
+are materialized.
+
+### Correctly Measuring Device Time (Compilation + Execution)
```

Comment:
nit: Compilation is part of the Host-Side Activity

Reply:
fixed. PTAL.